doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,192 @@
+tasks are executed on remote workers. You can find more details about `Dataset read execution here <https://docs.ray.io/en/master/data/key-concepts.html#reading-data>`__.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __mongo_datasource_reader_start__
+    :end-before: __mongo_datasource_reader_end__
+
+Now, we have finished implementing support for reading from a custom datasource!
+Let's move on to implementing support for writing back to the custom datasource.
+
+-------------
+Write support
+-------------
+
+Similar to read support, we start with handling a single block. Again 
+the ``PyMongo`` and  ``PyMongoArrow`` are used for MongoDB interactions.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __write_single_block_start__
+    :end-before: __write_single_block_end__
+
+Unlike read support, we do not need to implement a custom interface.
+
+Below, we implement a helper function to parallelize writing, which is
+expected to return a list of :ref:`Ray ObjectRefs <objects-in-ray>`. This helper function
+will later be used in the implementation of :meth:`~ray.data.Datasource.do_write`.
+
+In short, the below function spawns multiple :ref:`Ray remote tasks <ray-remote-functions>`
+and returns :ref:`their futures (object refs) <objects-in-ray>`.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __write_multiple_blocks_start__
+    :end-before: __write_multiple_blocks_end__
+
+-----------------------
+Putting it all together
+-----------------------
+
+With ``_MongoDatasourceReader`` and ``_write_multiple_blocks`` above, we are 
+ready to implement :meth:`create_reader() <ray.data.Datasource.create_reader>` 
+and :meth:`do_write() <ray.data.Datasource.do_write>`, and put together 
+a ``MongoDatasource``.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __mongo_datasource_start__
+    :end-before: __mongo_datasource_end__
+
+Now you can create a Ray Dataset from and write back to MongoDB, just like 
+any other datasource!
+
+.. code-block:: python
+
+    # Read from MongoDB datasource and create a dataset.
+    # The args are passed to MongoDatasource.create_reader().
+    ds = ray.data.read_datasource(
+        MongoDatasource(),
+        uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin",
+        database="my_db",
+        collection=="my_collection",
+        pipelines=my_pipelines, # See the example definition of ``my_pipelines`` above
+    )
+
+    # Data preprocessing with Dataset APIs here
+    # ...
+
+    # Write the dataset back to MongoDB datasource.
+    # The args are passed to MongoDatasource.do_write().
+    ds.write_datasource(
+        MongoDatasource(),
+        uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin",
+        database="my_db",
+        collection="my_collection"
+    )
```

Comment:
It would be great to have a section at the end, to encourage users to contribute their custom data sources back to OSS. Non-blocking and we can add in a followup.
Something like:

```
Contributions of custom datasource to Ray Datasets are welcome!
If you think your custom datasource implementation can benefit other users, 
we encourage you create GitHub issue on Ray to initiate discussion.
```